VS2017 building support

From the docs: `Visual Studio 2017 support has not been tested yet - waiting on node-gyp support to be generally available.`

Seems like `node-gyp` has this now: https://github.com/nodejs/node-gyp/commit/ae141e19060ecb1c596a96acf9510d66b7376c18

Any potential issues before I dive in on this?

Just doing another pass at this, and it seems we have to upgrade `npm` to 5.x to get a later version of `node-gyp` (`npm install -g node-gyp` doesn't seem to affect this):

```
> runas@3.1.1 install C:\Users\shiftkey\src\desktop\app\node_modules\runas
> node-gyp rebuild


C:\Users\shiftkey\src\desktop\app\node_modules\runas>if not defined npm_config_node_gyp (node "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
Traceback (most recent call last):
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\gyp_main.py", line 16, in <module>
    sys.exit(gyp.script_main())
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 538, in main
    return gyp_main(args)
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 514, in gyp_main
    options.duplicate_basename_check)
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 98, in Load
    generator.CalculateVariables(default_variables, params)
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py", line 1918, in CalculateVariables
    generator_flags.get('msvs_version', 'auto'))
  File "C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\MSVSVersion.py", line 434, in SelectVisualStudioVersion
    versions = _DetectVisualStudioVersions(version_map[version], 'e' in version)
KeyError: '2017'
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (C:\Users\shiftkey\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:336:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\shiftkey\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\shiftkey\src\desktop\app\node_modules\runas
gyp ERR! node -v v7.10.1
gyp ERR! node-gyp -v v3.6.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! runas@3.1.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the runas@3.1.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\shiftkey\AppData\Roaming\npm-cache\_logs\2017-10-16T02_21_01_802Z-debug.log
```

We're going to switch to `yarn` in #2864, but that too has this issue (it falls back to the system `npm` for `node-gyp`, rather than using a global `node-gyp`).

```
[4/4] Building fresh packages...
error C:\Users\shiftkey\src\desktop\app\node_modules\keytar: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\Users\shiftkey\src\desktop\app\node_modules\keytar
Output:
C:\Users\shiftkey\src\desktop\app\node_modules\keytar>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
gyp info it worked if it ends with ok
gyp info using node-gyp@3.5.0
gyp info using node@7.10.1 | win32 | x64
gyp info spawn C:\Python27\python.EXE
gyp info spawn args [ 'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'msvs',
gyp info spawn args   '-G',
gyp info spawn args   'msvs_version=2017',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Users\\shiftkey\\src\\desktop\\app\\node_modules\\keytar\\build\\config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Users\\shiftkey\\.node-gyp\\iojs-1.7.9\\common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=C:\\Users\\shiftkey\\.node-gyp\\iojs-1.7.9',
gyp info spawn args   '-Dnode_gyp_dir=C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp',
gyp info spawn args   '-Dnode_lib_file=iojs.lib',
gyp info spawn args   '-Dmodule_root_dir=C:\\Users\\shiftkey\\src\\desktop\\app\\node_modules\\keytar',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'C:\\Users\\shiftkey\\src\\desktop\\app\\node_modules\\keytar\\build',
gyp info spawn args   '-Goutput_dir=.' ]
Traceback (most recent call last):
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\gyp_main.py", line 16, in <module>
    sys.exit(gyp.script_main())
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 538, in main
    return gyp_main(args)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 514, in gyp_main
    options.duplicate_basename_check)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 98, in Load
    generator.CalculateVariables(default_variables, params)
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py", line 1900, in CalculateVariables
    generator_flags.get('msvs_version', 'auto'))
  File "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\MSVSVersion.py", line 434, in SelectVisualStudioVersion
    versions = _DetectVisualStudioVersions(version_map[version], 'e' in version)
KeyError: '2017'
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\configure.js:308:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\shiftkey\src\desktop\app\node_modules\keytar
gyp ERR! node -v v7.10.1
gyp ERR! node-gyp -v v3.5.0
gyp ERR! not ok
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

I think we'll wait on this until after we upgrade to Electron 1.8.x, which depends on Node 8.2.1 or later. I've started [the docs changes](https://github.com/desktop/desktop/compare/master...shiftkey-vs2017) here in the meantime to clarify the dependencies.

--- a/docs/contributing/setup.md
+++ b/docs/contributing/setup.md
@@ -18,11 +18,17 @@ You will need to install these tools on your machine:
     - *Let Python install into the default suggested path (`c:\Python27`), otherwise you'll have
       to configure node-gyp manually with the path which is annoying.*
     - *Ensure the **Add python.exe to Path** option is selected.*
- - Visual Studio 2015 or [Visual C++ Build Tools](http://go.microsoft.com/fwlink/?LinkId=691126)
-    - *If you already have Visual Studio 2015 installed, ensure you have the **Common Tools for Visual C++ 2015**
-      feature as that is required by Node.js for installing native modules.*
-    - *Visual Studio 2017 support has not been tested yet - see [#1766](https://github.com/desktop/desktop/issues/1766) for details*
- - *Run `npm config set msvs_version 2015` to tell node the right toolchain to use for compiling native modules.*
+ - One of Visual Studio 2015, Visual C++ Build Tools or Visual Studio 2017
+   - [Visual C++ Build Tools](http://go.microsoft.com/fwlink/?LinkId=691126)
+     - *Run `npm config set msvs_version 2015` to tell node to use this toolchain.*
+   - Visual Studio 2015 
+     - *Ensure you select the **Common Tools for Visual C++ 2015** feature as that is required by Node.js
+        for installing native modules.*
+     - *Run `npm config set msvs_version 2015` to tell node to use this toolchain.*
+   - [Visual Studio 2017](https://www.visualstudio.com/vs/community/)
+     - *Ensure you select the **Desktop development with C++** feature as that is required by Node.js for
+        installing native modules.*
+     - *Run `npm config set msvs_version 2017` to tell node to use this toolchain.*
 
 ### Fedora 26
 

